### PR TITLE
[FLINK-23912] Remove unnecessary "Clearing resource requirements of job" in SlotManager

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 1]
```diff
@@ -117,7 +117,7 @@ public void increaseResourceRequirementsBy(ResourceCounter increment) {
 
     @Override
     public void decreaseResourceRequirementsBy(ResourceCounter decrement) {
-        if (decrement.isEmpty()) {
+        if (decrement.isEmpty() || totalResourceRequirements.isEmpty()) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:120]: The new branch would hide a bug. There should never be a case where we decrease the resource requirement such that the result is negative. If such a case exists we should try to fix that instead.

Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:120]: IIUC, the negative requirement is by design not allowed. Could @zhuzhurk make sure about this?
The `ResourceCount` used here already covers this case.

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java (hunk 1 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:120]: I guess what @zentol means is that if we decrease a non-empty `ResourceCounter` when the `totalResourceRequirements` is not empty, it is an unexpected bug and needs to be exposed.
So I think we should add a check here and also fix `ResourceCounter#updateNewValue()` to throw exception if negative result happens.

Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:120]: Ahh. I think I misunderstand it.
> There should never be a case where we decrease the resource requirement such that the result is negative.
This scenario already exists, that is why we observe multiple "Clearing resource requirements of job". The `totalResourceRequirements` will decrease each time in `onReleaseTaskManager`. If that is a bug, we should fix it. @zentol

**Review by zentol — APPROVED** [review]
I rebased the PR and will merge it once CI is green.

**Pull request merged**